pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -118,10 +136,57 @@ func (cache *claimInfoCache) hasPodReference(UID types.UID) bool {
 	defer cache.RUnlock()
 
 	for _, claimInfo := range cache.claimInfo {
-		if claimInfo.podUIDs.Has(string(UID)) {
+		if claimInfo.PodUIDs.Has(string(UID)) {
 			return true
 		}
 	}
 
 	return false
 }
+
+func (cache *claimInfoCache) addPodReference(claimName, namespace string, podUID types.UID) error {
+	cache.Lock()
+	defer cache.Unlock()
+
+	claimInfo, exist := cache.claimInfo[claimName+namespace]
+	if !exist {
+		return fmt.Errorf("failed to get claimInfo with claim name %s and namespace %s", claimName, namespace)
+	}
+
+	claimInfo.PodUIDs.Insert(string(podUID))
+	return nil
+
+}
+
+func (cache *claimInfoCache) deletePodReference(claimName, namespace string, podUID types.UID) error {
+	cache.Lock()
+	defer cache.Unlock()
+
+	claimInfo, exist := cache.claimInfo[claimName+namespace]
+	if !exist {
+		return fmt.Errorf("failed to get claimInfo with claim name %s and namespace %s", claimName, namespace)
+	}
+
+	claimInfo.PodUIDs.Delete(string(podUID))
+	return nil
+}
+
+func (cache *claimInfoCache) syncToCheckpoint() error {
+	cache.RLock()
+
+	claimInfoStateList := make(state.ClaimInfoStateList, 0)
+	for _, infoClaim := range cache.claimInfo {
+		claimInfoState := state.ClaimInfoState{
+			DriverName: infoClaim.DriverName,
+			ClaimUID:   infoClaim.ClaimUID,
+			ClaimName:  infoClaim.ClaimName,
+			Namespace:  infoClaim.Namespace,
+			PodUIDs:    infoClaim.PodUIDs,
+			CdiDevices: infoClaim.CdiDevices,
+		}
+		claimInfoStateList = append(claimInfoStateList, claimInfoState)
+	}
```

Comment:
so in my code I am cloning the claimInfoStateList  so I can call cache.RUnlock(), but looking on you other comment it seem you prefer to RLock also the cache.state.Store call.

Reply:
I see. You made a copy so that you could unlock before writing to the store (which has its own lock). I don't have a huge preference here since then, except that if you are going to make a copy you should do a full DeepCopy.

Otherwise fields like PodUIDs and CdiDevices only have their slice pointers copied (and their contents could still subsequently change after the copy).